src/reaper/signal.rs

```diff
@@ -0,0 +1,238 @@
+//! A version of the reaper that waits for a signal to check for process progress.
+
+use async_lock::{Mutex as AsyncMutex, MutexGuard as AsyncMutexGuard};
+use event_listener::Event;
+use futures_lite::future;
+
+use std::io;
+use std::mem;
+use std::sync::Mutex;
+
+pub(crate) type Lock = AsyncMutexGuard<'static, ()>;
+
+/// The zombie process reaper.
+pub(crate) struct Reaper {
+    /// An event delivered every time the SIGCHLD signal occurs.
+    sigchld: Event,
+
+    /// The list of zombie processes.
+    zombies: Mutex<Vec<std::process::Child>>,
+
+    /// The pipe that delivers signal notifications.
+    pipe: Pipe,
+
+    /// Locking this mutex indicates that we are polling the SIGCHLD event.
+    driver_guard: AsyncMutex<()>,
+}
+
+impl Reaper {
+    /// Create a new reaper.
+    pub(crate) fn new() -> Self {
+        Reaper {
+            sigchld: Event::new(),
+            zombies: Mutex::new(Vec::new()),
+            pipe: Pipe::new().expect("cannot create SIGCHLD pipe"),
+            driver_guard: AsyncMutex::new(()),
+        }
+    }
+
+    /// Lock the driver thread.
+    pub(crate) async fn lock(&self) -> AsyncMutexGuard<'_, ()> {
+        self.driver_guard.lock().await
+    }
+
+    /// Reap zombie processes forever.
+    pub(crate) async fn reap(&'static self, _driver_guard: async_lock::MutexGuard<'_, ()>) -> ! {
+        loop {
+            // Wait for the next SIGCHLD signal.
+            self.pipe.wait().await;
+
+            // Notify all listeners waiting on the SIGCHLD event.
+            self.sigchld.notify(std::usize::MAX);
+
+            // Reap zombie processes, but make sure we don't hold onto the lock for too long!
+            let mut zombies = mem::take(&mut *self.zombies.lock().unwrap());
+            let mut i = 0;
+            'reap_zombies: loop {
+                for _ in 0..50 {
+                    if i >= zombies.len() {
+                        break 'reap_zombies;
+                    }
+
+                    if let Ok(None) = zombies[i].try_wait() {
+                        i += 1;
+                    } else {
+                        zombies.swap_remove(i);
+                    }
+                }
+
+                // Be a good citizen; yield if there are a lot of processes.
+                //
+                // After we yield, check if there are more zombie processes.
+                future::yield_now().await;
+                zombies.append(&mut self.zombies.lock().unwrap());
```

Comment:
maybe put this at the top of the loop (and initialize `zombies` with an `Vec::new()`)

Reply:
Taking out the allocation of `zombies` saves us an allocation for the temporary buffer.